service/history/historyEngine.go

```diff
@@ -2918,48 +2996,67 @@ func (e *historyEngineImpl) applyWorkflowIDReusePolicyForSignalWithStart(
 		execution,
 		wfIDReusePolicy,
 	)
-
 }
 
 func (e *historyEngineImpl) applyWorkflowIDReusePolicyHelper(
 	prevStartRequestID,
 	prevRunID string,
 	prevState enumsspb.WorkflowExecutionState,
 	prevStatus enumspb.WorkflowExecutionStatus,
-	execution commonpb.WorkflowExecution,
+	newExecution commonpb.WorkflowExecution,
 	wfIDReusePolicy enumspb.WorkflowIdReusePolicy,
-) error {
+) (prevExecutionUpdateFn updateWorkflowActionFunc, err error) {
 
 	// here we know there is some information about the prev workflow, i.e. either running right now
 	// or has history check if the this workflow is finished
 	switch prevState {
 	case enumsspb.WORKFLOW_EXECUTION_STATE_CREATED,
 		enumsspb.WORKFLOW_EXECUTION_STATE_RUNNING:
+		if wfIDReusePolicy == enumspb.WORKFLOW_ID_REUSE_POLICY_TERMINATE_IF_RUNNING {
+			return func(context workflow.Context, mutableState workflow.MutableState) (*updateWorkflowAction, error) {
+				if !mutableState.IsWorkflowExecutionRunning() {
+					// found workflow already closed after loading, nothing we need to do
+					return updateWorkflowWithoutWorkflowTask, nil
```

Comment:
I should return an error in this case and fall back to just creating the new workflow.